### ISSUE TITLE
Multiple categories of routes

### ISSUE THREAD

**Issue by ShadesOT** [We-the-People-civ4col-mod/Mod#188]
It would be nice if the mod could handle multiple categories of routes on a plot.
Eg
```
Category "road"
tier 1: road
tier 2: country road
tier 3: plastered country road

Category "railroad track"
tier 1: basic track
tier 2: fast track

Categroy "steampunk futuristic parallel world transportation system. useable by medics only"
tier 1:  whatever
```
This is just to draw an example.

- Categories and tiers can be unlocked by CivEffects.
- CivEffects can limit units to the use of certain categories.
- CivEffects can define which unit type can build which category of transportation improvement.
- Improvements in different categories could and would need to be built independently.

I am proposing this with the return of railroads ( #98 ) in mind as the primary motivation. Then roads and rail track would be sepparate categories. Trains could only use the rail track category. All other units could only use the road category. Pioneers would be able to build routes in the road category, the new navvy unit could build routes in both categories. Something along those lines. Doesn't have to be exactly like that. 'Just to illustrate the possibilities.

**Comment by ShadesOT** [comment]
> As for the idea of having route categories. I'm not sure I like it. What concerns me is that it will be fairly complex, particularly regarding unit movement, which is one of the most critical parts of the game regarding performance.

How does this make routing more complex? If a train can only move on plots with rail track routes, I would think the pathfinding becomes simpler.

> We also have the issue that the graphics engine can only display one route on each plot.

That is an issue I think ...

> CivEffects can enable/disable builds.

What do you mean by "builds" in this context?

> I guess you could make two builds, which are identical, make one profession use one and another profession use another. That way you can enable building the same route for two different professions with two different CivEffects.

What is the issue? Can we not use a bit of entity-relationship structure in the civeffects xml structure to achieve this easily? How you describe the structure, it is doing that already.
```
routes:
  category_road
  category_railtrack

builds:
  buildA:
    canBuildRouteCategory:
      category_road
    canBuildImprovements:
      ...
  buildB:
    canBuildRouteCategory:
      category_road
      category_railtrack
    canBuildImprovements:
      none

units:
  unit_pioneer:
    buildA
  unit_navvy:
    buildB
```



> we could skip improvements (farms etc) for navvies, meaning they become master route builders, but can't do anything but routes.

I like that. 👍 

> At the very least we should do something like:
> ```c++
> int iMinRoute = this->getMinRoute();
> if (iMinRoute > NO_ROUTE && iMinRoute < plot()->getRoute()) return false;
> ```
> 
> This is in CvUnit::canMoveInto (or whatever it's called). getMinRoute is then from CvUnitInfo, which sets it as an xml tag. If the unit xml entry lags the tag, it defaults to NO_ROUTE.

In case we do not implement categories, how about we do it like this:
```cpp
enum RouteTypes
{
  NO_ROUTE = -1,
  ROAD,
  COUNTRY_ROAD,
  ROAD_AND_TRACK,
  COUNTRY_ROAD_AND_TRACK,
  TRACK
};

int iMinRoute = this->getMinRoute();
int iMaxRoute = this->getMaxRoute();
int iPlotRoute = plot()->getRoute();
if ( iPlotRoute >= iMinRoute && iPlotRoute <= iMaxRoute)
{
  //can move on plot
}
```
Ugly, and does not scale, but it would work for our purposes. This way normal units can not use track only plots. And we can make graphics for all 5 route types.

**Comment by Nightinggale** [comment]
`
How does this make routing more complex? 
`
There are two parts to pathfinding. One is canMoveInto. The other one is movement cost. Both are called a lot of times each time a unit considers to move. Adding complexity to movement cost will slow down finding the path with the lowest cost.

`
What do you mean by "builds" in this context?
`
xml/units/CIV4BuildInfos.xml. The pioneer actions are all written in this file. It sets stuff like build time, which route/improvement/feature/terrain the plot will end up with and so on. This means it's a description of the build action while the other files describe the results.

```cpp
int iMinRoute = this->getMinRoute();
int iMaxRoute = this->getMaxRoute();
```

After having thought about it, unitinfo should have two arrays. One is which routes it can use and one is which routes are required. Those are then stored in a single unsigned int when reading xml.

During movement, the code should be something like
```cpp
if (!(unitinfo->getRouteRequirements() & ( 1 << (pPlot->getRoute() + 1)))) return false;
```
What happens is that each route gets a single bit in the int, like the unsigned int is used as a bool array with 32 entries. It looks up entry getRoute()+1. It's offset by 1 to add support for NO_ROUTE, which is actually -1. In other words it's like writing:
```cpp
bool[32] routeRequirements;
if (!routeRequirements[pPlot->getRoute() + 1]) return false;
```
What is clever about the unsigned int approach is that it keeps the memory local. It saves a pointer lookup relative to using a normal array or vector. It uses and, bitshift, binary and, each of those takes 1 CPU cycle and some CPUs can do more than one of those during a single cycle and the check becomes perfect for a high performance function.

CivEffect has introduced the function HasBit, which makes the function call something like this:
```cpp
if (!HasBit(unitinfo->getRouteRequirements(), pPlot->getRoute() + 1)) return false;
```
Same thing, but hopefully with a bit more readable code.

The price of this approach is a hard limit of 31 different route types, but using more than 31 seems bloated to me anyway.

**Comment by ShadesOT** [comment]
If NO_ROUTE would be 0 instead of -1, we could even save that one addition all the time.

Bit masking ... I really like it. We should use it much more often throughout the code.

**Comment by Nightinggale** [comment]
`
If NO_ROUTE would be 0 instead of -1, we could even save that one addition all the time.
`
The standard is that NO_ROUTE is -1 while the rest are xml indexes. This means we can do GC.getRouteInfo(pPlot->getRoute()). It's not exclusive to routes. All xml files work like that.

Leaving the standard will be prone to bugs everywhere else. Adding one addition of a fixed number is really a 0-1 cycle operation and isn't something we should be concerned about. We can do this thousands of times for the time it takes to read an int from memory if it's a cache miss. It's the wrong place to optimize.

`
Bit masking ... I really like it. We should use it much more often throughout the code.
`
M:C has bitmasks in multiple info classes. It's useful for multiple reasons:

- It saves memory (0 bytes for the first, 4 bytes for each of the following 31 bools)
- It uses the same 32 bit for multiple bool lookups, meaning if multiple bools have to be checked, they are in the CPU cache
- Bitmasks can make or checks fast (like bitmask for checking isLightArmor() || isHeavyArmor() in one cycle)

The saved memory doesn't matter in general, but it's an ok bonus. "Compressing" the bools to have more of them in L1 cache at once is likely the most important reason to take this path. Bitmask checking for multiple bool checks can be used, but it's certainly not the most common use case.

The "key" to reading the bools from such a variable should be set in an enum. Otherwise it becomes unreadable and prone to bugs.

BoolArray (CivEffect branch) is essentially a bitfield. It's an array of unsigned int32 where it stores 32 bools in each. If the length of the array is less than 32, it will not allocate memory, but rather use the pointer memory as the array, preventing the need to resolve the pointer in memory. It can be accessed like JIT arrays, meaning it has get and set functions as well as read and write for savegames.

**Comment by ShadesOT** [comment]
> BoolArray (CivEffect branch) is essentially a bitfield. It's an array of unsigned int32 where it stores 32 bools in each. If the length of the array is less than 32, it will not allocate memory, but rather use the pointer memory as the array, preventing the need to resolve the pointer in memory.

Great! Did you overload operators? I would like it to be able to do this:
```cpp
//bitmasks
uint bm1 = 1;
uint bm2 = 3;

BoolArray bArray1 = new BoolArray(bm1);
BoolArray bArray2 = new BoolArray(bm2);
BoolArray bArray3 = new BoolArray(bArray1 & bArray2); //should be equal to bm1

//or even better:
bArray3 = bArray1 & bArray2
```

**Comment by ErikSandbraaten** [comment]
There's already 3 alternative bitset implementations
boost::dynamic_bitset, std::bitset and vector<bool>.
Pardon my ignorance, but why are we inventing a new one ?

**Comment by Nightinggale** [comment]
`
Did you overload operators? 
`
No, but we can add overloaded operators if we find something useful.

`
Pardon my ignorance, but why are we inventing a new one ?
`
Because I reuse the implementation I used in the last project, where I reused it from the previous project.... and eventually dating back to before I touched boost. In other words the HasBit and SetBit I added are older than Colonization itself and has worked flawlessly the entire time. It's not a new untested addition.

As for BoolArray, it has added functions like savegames and working together with the other array classes I added in the CivEffect branch. Also BoolArray takes an enum value as argument to the constructor, which locks it to a certain xml file, meaning it automatically gets the length of say NUM_YIELD_TYPES. It's aware of this and will assert if used as arguments for something, which expects something else like promotions.
